python/ray/data/_internal/arrow_serialization.py

```diff
@@ -88,31 +112,475 @@ def _register_arrow_data_serializer(serialization_context):
+
+    # Reduce the ChunkedArray columns.
+    reduced_columns = []
+    for column_name in t.column_names:
+        column = t[column_name]
+        try:
+            # Delegate to ChunkedArray reducer.
+            reduced_column = _arrow_chunked_array_reduce(column)
+        except Exception as e:
+            if not _is_dense_union(column.type) and _is_in_test():
+                # If running in a test and the column is not a dense union array
+                # (which we expect to need a fallback), we want to raise the error,
+                # not fall back.
+                raise e from None
+            if type(column.type) not in _serialization_fallback_set:
+                logger.warning(
+                    "Failed to complete optimized serialization of Arrow Table, "
+                    f"serialization of column '{column_name}' of type {column.type} "
+                    "failed, so we're falling back to Arrow IPC serialization for the "
+                    "table. Note that this may result in slower serialization and more "
+                    "worker memory utilization. Serialization error:",
+                    exc_info=True,
+                )
+                _serialization_fallback_set.add(type(column.type))
+            # Fall back to Arrow IPC-based workaround for the entire table.
+            return _arrow_table_ipc_reduce(t)
+        else:
+            # Column reducer succeeded, add reduced column to list.
+            reduced_columns.append(reduced_column)
+    return _reconstruct_table, (reduced_columns, t.schema)
+
+
+def _reconstruct_table(
+    reduced_columns: List[Tuple[List["pyarrow.Array"], "pyarrow.DataType"]],
+    schema: "pyarrow.Schema",
+) -> "pyarrow.Table":
+    """Restore a serialized Arrow Table, reconstructing each reduced column."""
+    import pyarrow as pa
+
+    # Reconstruct each reduced column.
+    columns = []
+    for chunks, type_ in reduced_columns:
+        columns.append(_reconstruct_chunked_array(chunks, type_))
+
+    return pa.Table.from_arrays(columns, schema=schema)
+
+
+def _arrow_chunked_array_reduce(
+    ca: "pyarrow.ChunkedArray",
+) -> Tuple[List["pyarrow.Array"], "pyarrow.DataType"]:
+    """Custom reducer for Arrow ChunkedArrays that works around a zero-copy slice
+    pickling bug. This reducer does not return a reconstruction function, since it's
+    expected to be reconstructed by the Arrow Table reconstructor.
+    """
+    truncated_chunks = []
+    for chunk in ca.chunks:
+        chunk = _copy_array_if_needed(chunk)
+        truncated_chunks.append(chunk)
+    return truncated_chunks, ca.type
+
+
+def _reconstruct_chunked_array(
+    chunks: List["pyarrow.Array"], type_: "pyarrow.DataType"
+) -> "pyarrow.ChunkedArray":
+    """Restore a serialized Arrow ChunkedArray from chunks and type."""
+    import pyarrow as pa
+
+    return pa.chunked_array(chunks, type_)
+
+
+def _copy_array_if_needed(a: "pyarrow.Array") -> "pyarrow.Array":
```

Comment:
Can test _copy_array_if_needed, _copy_array_buffers_if_needed? They are dealing with more heavy logic.

Reply:
`_copy_array_if_needed` is essentially covered by https://github.com/ray-project/ray/blob/9d9ca20efba3f20f7622148458662c04c729d9ba/python/ray/data/tests/test_arrow_serialization.py#L152-L359, since `_arrow_table_reduce` and `_arrow_chunked_array_reduce` are both thin layers, and the end effect of propagating array-level slices to buffer-level slices is a bit difficult to observe other than in the eventual pickle payload size... I can duplicate the e2e pickling tests with some array-in, array-out tests that assert that a copy is made in certain situations and not made in others, but that's probably going to be another ~500 lines of tests.

I'll try to find a better way to structure this so it's more concisely testable, but to derisk this landing for the branch cut, can you also review this as-is?